File: app/application/repositories/snapshot.py
"""Snapshot repository for persistence of Snapshot entities.

Provides date-ordered queries, latest-by-camera lookups, and retention
cleanup over an async SQLAlchemy session.
"""

from datetime import date, datetime

from sqlalchemy import select, func, delete, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.domain.models import Snapshot


class SnapshotRepository:
    """Provide query and mutation access to Snapshot records.

    All mutation methods flush the session so callers can rely on
    generated columns and rowcount metadata immediately afterwards.
    """

    def __init__(self, session: AsyncSession):
        """Bind the repository to the supplied async session.

        Args:
            session: The async SQLAlchemy session used for queries.
        """
        self._session = session

    async def get_by_id(self, snapshot_id: int) -> Snapshot | None:
        """Get a single snapshot by its primary key.

        Args:
            snapshot_id: The unique identifier of the snapshot.

        Returns:
            The matching Snapshot, or None when not found.
        """
        result = await self._session.execute(
            select(Snapshot).where(Snapshot.id == snapshot_id)
        )
        return result.scalar_one_or_none()

    async def get_by_camera_and_date(
        self, camera_id: int, target_date: date
    ) -> list[Snapshot]:
        """Get all snapshots for a camera captured on a given date.

        Args:
            camera_id: The identifier of the camera.
            target_date: The calendar date to query (full day range).

        Returns:
            Snapshots ordered by captured_at timestamp.
        """
        start = datetime.combine(target_date, datetime.min.time())
        end = datetime.combine(target_date, datetime.max.time())
        result = await self._session.execute(
            select(Snapshot)
            .where(
                Snapshot.camera_id == camera_id,
                Snapshot.captured_at >= start,
                Snapshot.captured_at <= end,
            )
            .order_by(Snapshot.captured_at)
        )
        return list(result.scalars().all())

    async def get_by_date(self, target_date: date) -> list[Snapshot]:
        """Get all snapshots captured on a given date across cameras.

        Args:
            target_date: The calendar date to query (full day range).

        Returns:
            Snapshots ordered by camera_id then captured_at timestamp.
        """
        start = datetime.combine(target_date, datetime.min.time())
        end = datetime.combine(target_date, datetime.max.time())
        result = await self._session.execute(
            select(Snapshot)
            .where(Snapshot.captured_at >= start, Snapshot.captured_at <= end)
            .order_by(Snapshot.camera_id, Snapshot.captured_at)
        )
        return list(result.scalars().all())

    async def get_last_by_camera(self, camera_id: int) -> Snapshot | None:
        """Get the most recent successful snapshot for a camera.

        Args:
            camera_id: The identifier of the camera.

        Returns:
            The latest successful Snapshot, or None when none exist.
        """
        result = await self._session.execute(
            select(Snapshot)
            .where(Snapshot.camera_id == camera_id, Snapshot.status == "success")
            .order_by(Snapshot.captured_at.desc())
            .limit(1)
        )
        return result.scalar_one_or_none()

    async def get_last_for_all_cameras(
        self, camera_ids: list[int]
    ) -> dict[int, Snapshot | None]:
        """Get the latest successful snapshot for each given camera.

        Args:
            camera_ids: List of camera identifiers to look up.

        Returns:
            Dict keyed by camera_id with the latest successful Snapshot
            or None when a camera has no successful snapshots.
        """
        if not camera_ids:
            return {}
        result = await self._session.execute(
            select(Snapshot)
            .where(
                Snapshot.camera_id.in_(camera_ids),
                Snapshot.status == "success",
            )
            .order_by(Snapshot.captured_at.desc())
        )
        rows = result.scalars().all()
        seen: set[int] = set()
        latest: dict[int, Snapshot | None] = {cid: None for cid in camera_ids}
        for row in rows:
            if row.camera_id not in seen:
                seen.add(row.camera_id)
                latest[row.camera_id] = row
        return latest

    async def add(self, snapshot: Snapshot) -> None:
        """Stage a new Snapshot and flush it to the session.

        Args:
            snapshot: The Snapshot instance to persist.
        """
        self._session.add(snapshot)
        await self._session.flush()

    async def get_all_successful_with_images(self) -> list[Snapshot]:
        """Get all successful snapshots that have an image file saved.

        Used by the live manifest endpoint to build the per-date
        snapshot index without a separate export step.

        Returns:
            Snapshots ordered by camera_id then captured_at timestamp.
        """
        result = await self._session.execute(
            select(Snapshot)
            .where(
                Snapshot.status == "success",
                Snapshot.image_path != "",
            )
            .order_by(Snapshot.camera_id, Snapshot.captured_at)
        )
        return list(result.scalars().all())

    async def count_by_camera_and_date(
        self, camera_id: int, target_date: date
    ) -> int:
        """Count snapshots for a camera on a given date.

        Args:
            camera_id: The identifier of the camera.
            target_date: The calendar date to query (full day range).

        Returns:
            The number of matching snapshots, or 0 when none exist.
        """
        start = datetime.combine(target_date, datetime.min.time())
        end = datetime.combine(target_date, datetime.max.time())
        result = await self._session.execute(
            select(func.count(Snapshot.id)).where(
                Snapshot.camera_id == camera_id,
                Snapshot.captured_at >= start,
                Snapshot.captured_at <= end,
            )
        )
        return result.scalar() or 0

    async def delete_older_than(self, cutoff: datetime) -> int:
        """Delete snapshots captured before a cutoff timestamp.

        Args:
            cutoff: The exclusive lower-bound timestamp; older rows are removed.

        Returns:
            The number of snapshots deleted.
        """
        result = await self._session.execute(
            delete(Snapshot).where(Snapshot.captured_at < cutoff)
        )
        await self._session.flush()
        return result.rowcount

    async def get_old_unarchived(self, cutoff: datetime) -> list[Snapshot]:
        """Get snapshots older than cutoff that haven't been archived yet.

        Args:
            cutoff: The exclusive lower-bound timestamp.

        Returns:
            Snapshots ordered by camera_id then captured_at.
        """
        result = await self._session.execute(
            select(Snapshot)
            .where(Snapshot.captured_at < cutoff, Snapshot.archive_path.is_(None))
            .order_by(Snapshot.camera_id, Snapshot.captured_at)
        )
        return list(result.scalars().all())

    async def update_archive_path(self, snapshot_id: int, archive_path: str) -> None:
        """Set the archive_path for a snapshot.

        Args:
            snapshot_id: The identifier of the snapshot.
            archive_path: The archive path to store (e.g. ``snapshots/3/2025-06-15.zip::150322.jpg``).
        """
        await self._session.execute(
            update(Snapshot)
            .where(Snapshot.id == snapshot_id)
            .values(archive_path=archive_path)
        )
        await self._session.flush()

    async def count_by_archive_zip(self, zip_path: str) -> int:
        """Count snapshots referencing a given archive zip.

        Args:
            zip_path: The archive zip path to match (e.g. ``snapshots/3/2025-06-15.zip``).

        Returns:
            The number of snapshots that reference this zip.
        """
        result = await self._session.execute(
            select(func.count(Snapshot.id)).where(
                Snapshot.archive_path.startswith(zip_path)
            )
        )
        return result.scalar() or 0

File: internal/repository/snapshot.go
package repository

import (
	"context"
	"time"

	"github.com/jmoiron/sqlx"

	"github.com/Lmex89/home-cameras/internal/domain"
)

// SnapshotRepository persists Snapshot entities.
type SnapshotRepository struct {
	db DBTX
}

// NewSnapshotRepository builds a snapshot repository bound to db.
func NewSnapshotRepository(db DBTX) *SnapshotRepository { return &SnapshotRepository{db: db} }

const snapshotCols = `id, camera_id, captured_at, image_path, file_size, status,
	error_message, archive_path`

// GetByID returns a snapshot or an error when missing.
func (r *SnapshotRepository) GetByID(ctx context.Context, id int64) (*domain.Snapshot, error) {
	var s domain.Snapshot
	err := sqlx.GetContext(ctx, r.db, &s,
		"SELECT "+snapshotCols+" FROM snapshots WHERE id = ?", id)
	if err != nil {
		return nil, err
	}
	return &s, nil
}

// GetByCameraAndDate returns a camera's snapshots for one full day.
func (r *SnapshotRepository) GetByCameraAndDate(ctx context.Context, cameraID int64, day time.Time) ([]domain.Snapshot, error) {
	start := dayInLocation(day, true)
	end := dayInLocation(day, false)
	var snaps []domain.Snapshot
	err := sqlx.SelectContext(ctx, r.db, &snaps,
		"SELECT "+snapshotCols+` FROM snapshots
		 WHERE camera_id = ? AND captured_at >= ? AND captured_at <= ?
		 ORDER BY captured_at`, cameraID, start, end)
	return snaps, err
}

// GetByDate returns all snapshots across cameras for one full day.
func (r *SnapshotRepository) GetByDate(ctx context.Context, day time.Time) ([]domain.Snapshot, error) {
	start := dayInLocation(day, true)
	end := dayInLocation(day, false)
	var snaps []domain.Snapshot
	err := sqlx.SelectContext(ctx, r.db, &snaps,
		"SELECT "+snapshotCols+` FROM snapshots
		 WHERE captured_at >= ? AND captured_at <= ?
		 ORDER BY camera_id, captured_at`, start, end)
	return snaps, err
}

// GetLastByCamera returns the most recent successful snapshot.
func (r *SnapshotRepository) GetLastByCamera(ctx context.Context, cameraID int64) (*domain.Snapshot, error) {
	var s domain.Snapshot
	err := sqlx.GetContext(ctx, r.db, &s,
		"SELECT "+snapshotCols+` FROM snapshots
		 WHERE camera_id = ? AND status = 'success'
		 ORDER BY captured_at DESC LIMIT 1`, cameraID)
	if err != nil {
		return nil, err
	}
	return &s, nil
}

// GetLastForAllCameras returns the latest successful snapshot per camera.
func (r *SnapshotRepository) GetLastForAllCameras(ctx context.Context, cameraIDs []int64) (map[int64]*domain.Snapshot, error) {
	out := make(map[int64]*domain.Snapshot, len(cameraIDs))
	if len(cameraIDs) == 0 {
		return out, nil
	}
	query, args, err := sqlx.In(
		"SELECT "+snapshotCols+` FROM snapshots
		 WHERE camera_id IN (?) AND status = 'success'
		 ORDER BY captured_at DESC`, cameraIDs)
	if err != nil {
		return nil, err
	}
	var snaps []domain.Snapshot
	if err := sqlx.SelectContext(ctx, r.db, &snaps, r.db.Rebind(query), args...); err != nil {
		return nil, err
	}
	seen := map[int64]bool{}
	for _, s := range snaps {
		if !seen[s.CameraID] {
			seen[s.CameraID] = true
			cp := s
			out[s.CameraID] = &cp
		}
	}
	return out, nil
}

// Add inserts a snapshot and populates its generated id.
func (r *SnapshotRepository) Add(ctx context.Context, s *domain.Snapshot) error {
	res, err := r.db.ExecContext(ctx,
		`INSERT INTO snapshots (camera_id, captured_at, image_path, file_size, status, error_message)
		 VALUES (?, ?, ?, ?, ?, ?)`,
		s.CameraID, s.CapturedAt, s.ImagePath, s.FileSize, s.Status, s.ErrorMessage)
	if err != nil {
		return err
	}
	id, err := res.LastInsertId()
	if err != nil {
		return err
	}
	s.ID = id
	return nil
}

// GetAllSuccessfulWithImages returns successful snapshots with files.
func (r *SnapshotRepository) GetAllSuccessfulWithImages(ctx context.Context) ([]domain.Snapshot, error) {
	var snaps []domain.Snapshot
	err := sqlx.SelectContext(ctx, r.db, &snaps,
		"SELECT "+snapshotCols+` FROM snapshots
		 WHERE status = 'success' AND image_path != ''
		 ORDER BY camera_id, captured_at`)
	return snaps, err
}

// DeleteOlderThan deletes snapshots captured before cutoff.
func (r *SnapshotRepository) DeleteOlderThan(ctx context.Context, cutoff time.Time) (int64, error) {
	res, err := r.db.ExecContext(ctx,
		"DELETE FROM snapshots WHERE captured_at < ?", cutoff.Format("2006-01-02 15:04:05"))
	if err != nil {
		return 0, err
	}
	n, err := res.RowsAffected()
	return n, err
}

// GetOldUnarchived returns snapshots older than cutoff without archives.
func (r *SnapshotRepository) GetOldUnarchived(ctx context.Context, cutoff time.Time) ([]domain.Snapshot, error) {
	var snaps []domain.Snapshot
	err := sqlx.SelectContext(ctx, r.db, &snaps,
		"SELECT "+snapshotCols+` FROM snapshots
		 WHERE captured_at < ? AND archive_path IS NULL
		 ORDER BY camera_id, captured_at`, cutoff.Format("2006-01-02 15:04:05"))
	return snaps, err
}

// UpdateArchivePath sets the archive reference for a snapshot.
func (r *SnapshotRepository) UpdateArchivePath(ctx context.Context, id int64, ref string) error {
	_, err := r.db.ExecContext(ctx,
		"UPDATE snapshots SET archive_path = ? WHERE id = ?", ref, id)
	return err
}

// MarkArchivedBatch sets the same archive ref on many snapshots.
func (r *SnapshotRepository) MarkArchivedBatch(ctx context.Context, ids []int64, ref string) (int64, error) {
	if len(ids) == 0 {
		return 0, nil
	}
	query, args, err := sqlx.In(
		"UPDATE snapshots SET archive_path = ? WHERE id IN (?)", ref, ids)
	if err != nil {
		return 0, err
	}
	res, err := r.db.ExecContext(ctx, r.db.Rebind(query), args...)
	if err != nil {
		return 0, err
	}
	return res.RowsAffected()
}

// CountByArchiveZip counts snapshots referencing an archive prefix.
func (r *SnapshotRepository) CountByArchiveZip(ctx context.Context, zipPath string) (int, error) {
	var n int
	err := sqlx.GetContext(ctx, r.db, &n,
		"SELECT COUNT(id) FROM snapshots WHERE archive_path LIKE ?", zipPath+"%")
	return n, err
}

// dayInLocation returns the start (or exclusive end) of a day.
func dayInLocation(day time.Time, start bool) string {
	y, m, d := day.Date()
	t := time.Date(y, m, d, 0, 0, 0, 0, day.Location())
	if !start {
		t = t.AddDate(0, 0, 1)
	}
	return t.Format("2006-01-02 15:04:05")
}
